### ISSUE TITLE
Prepare and review Design Document

### ISSUE THREAD

**Issue by loleg** [frictionlessdata/tableschema-jl#4]
As in the other libraries (e.g. [Go](https://github.com/frictionlessdata/tableschema-go/issues/4)), a design document helps to provide an overview of the functions and API calls provided in the library. Besides aggregating information about the work to be done, implementation and testing the document should contain a proposal for the first version of the API. The idea is to make it easier to discuss and gather comments. Have someone more acquainted to the libraries (and with experience in statically typed languages) signing off the document would be great.

**Comment by loleg** [comment]
The design document has been [drafted here](https://github.com/loleg/tableschema-jl/blob/master/DESIGN.md) and is open to discussion. Notable differences from other languages:

- Package and module names are CamelCase
- Project repositories have a dot extension instead of a dash, i.e. this project should live at https://github.com/frictionlessdata/TableSchema.jl to be conformant with package guidelines
- Code and tool links updated as appropriate.

**Comment by roll** [comment]
@loleg 
I'll be able to take a closer look tomorrow but it will be great to have an API as close as possible to https://github.com/frictionlessdata/implementations#tableschema (Python/JavaScript implementations are 100% compliant to it)

**Comment by roll** [comment]
@loleg 
It looks good and as said everyone will benefit from being close to Python/JavaScript API (users, maintainers, tutorial writers etc). A few comments:

```jl
filestream = os.open("schema.json")
schema = TableSchema.read(filestream) # you mean Schema.read? Also Schema.read("schema.json")?

err = schema.errors
# err is falsy, else an error summary

# we use
# schema.errors: Array
# schema.valid: Boolean

filestream = os.open("data.csv")
table = TableSchema.read(filestream) # you mean Table.read? Also Table.read("data.json")? And we have to accepts optional schema argument
err = table.errors

records = [ Dict(
    "name" => "header:name",
    "age" => "header:age"
) ]
filter = table.filter(records) # filter?
err = filter.errors

println("$i - $r") for (i, r) in filter
```

---

Related to repo renaming - is it mandatory for Julia project? If it's just a recommendation I think we better preserve an organisation style of naming. But anyway it's totally minor.

**Comment by loleg** [comment]
Thanks for the feedback. In addition to `Table.read`, should we also support constructors with these arguments? 

`table.filter` is coming from the Go design document I adapted. I'm also confused about where that is coming from, and have removed it. 

What parts of the API are deemed important to document at this stage? I have been trying to think ahead this week and look at several scenarios for what happens to the data once loaded. 

Often the best way to have a clear vision is just to work towards an alpha release, in my opinion -
 especially since from a design perspective the groundwork has already been laid for some time.

**Comment by loleg** [comment]
@roll I've updated the mock code in the [design document](https://github.com/loleg/tableschema-jl/blob/master/DESIGN.md). I don't have the API working exactly this way yet, but it's doable.

**Comment by roll** [comment]
@loleg 
Just realized that we can't use `Schema/Table.read` as factory methods (if I've got it right) because methods with the same name is used for data reading - `table.read()`. In JavaScript we user `Schema.load` and `Table.load` as constructors.

About API documentation I don't think it's critical for now. Again based on my understanding Julia is very close to Python so I think we better just highlight for now discrepancies with:
- https://github.com/frictionlessdata/implementations#interface
- https://github.com/frictionlessdata/tableschema-py#tableschema-py

**Comment by loleg** [comment]
Thanks, will revise.

**Comment by loleg** [comment]
In the implementation notes I see now the constructor and `table.read()`, so have adjusted design & spec accordingly.

**Comment by loleg** [comment]
In Julia, CSV files are usually read either using the [Base](https://docs.julialang.org/en/stable/stdlib/io-network/#Base.DataFmt.readcsv) function `readcsv` to output an `Array`, or the [CSV.jl](https://github.com/JuliaData/CSV.jl) library, whose `CSV.read()` call by default creates a [DataFrames.jl](http://juliadata.github.io/DataFrames.jl/stable/man/getting_started.html#Importing-and-Exporting-Data-(I/O)-1) object. Should we support both, or focus on one for now? I've written some initial tests using the former in the previous commit.

**Comment by roll** [comment]
@loleg 
In other language our intention is not relaying on other libraries data structures. E.g. if in JavaScript we could use native `Data` object to output data it's better than to use `externalLib.Data`. There is exceptions but overall direction is this.

Other question may be this `DataFrame` object compatible to `Array`. So user could use it as an array. But anyway I'm not really able to help more with Julia-specific things.

**Comment by roll** [comment]
Choosing the csv parser lib one of the most important thing is to check streaming capabilities.

**Comment by loleg** [comment]
@roll library choice is implementation detail. I am looking into [tabulator](https://pypi.python.org/pypi/tabulator/1.3.0), and we should pursue this discussion in #11.

I don't see any major API differences right now, but I can keep the Design doc up to date as they come up. Could you let me know if there is any more specific details required to wrap this up?

**Comment by loleg** [comment]
@roll one difference I ran into is that I'm unable to use `type` as field name since it is reserved. For now I'm using `of_type` as workaround. Any other preferences?

**Comment by roll** [comment]
@loleg 
I think we could use some Julia idiom for this case (may be it's `of_type`). On top of my head are only `data_type` also (probably in pair with `data_format` renaming).

**Comment by loleg** [comment]
Another alternative would be to access the dictionary directly, if API
translation is not required at that level. But I will look for comparable
idioms, thanks.

**Comment by loleg** [comment]
In a not completely systematic or exhaustive, but I think sufficient search
across Julia packages I found `_type` to be used on several occasions, and
have refactored accordingly.


Predominant is usage of `someType` and `typeSomething`, but not clearly
applicable here. I also realised that `of_type` can be confused with a
method call or something related to the `typeof` keyword.


Finally, I considered `:type` and alternative cases, but clarity is
lacking. In that sense my preference would be `datatype` or `columnType`.

**Comment by loleg** [comment]
@roll in a not completely systematic or exhaustive search across a few dozen Julia packages I found `_type` to be used on several occasions. 

Predominant is usage of `someType` and `typeSomething`, but not clearly applicable here. I also realised that `of_type` can be confused with a method call or something related to the `typeof` keyword.

I also considered `:type` and alternative cases, but clarity is lacking. In that sense my preference would be `datatype` or `columnType`. To keep things short and readable, what about `typed`, as in:

```
a_field.name = "A field"
a_field.typed = "integer"
```

Another alternative would be to access the dictionary directly, if API translation were not required at that level.
